packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+                }
+            }
+        }
+    }
+
+    void checkExportPackage(core::MutableContext &ctx, core::LocOffsets loc, core::NameRef litPackage) {
+        if (litPackage != this->package.mangledName() && !this->package.importsPackage(litPackage)) {
+            if (auto e = ctx.beginError(loc, core::errors::Packager::InvalidExport)) {
+                e.setHeader("Exporting a symbol that isn't owned by this package");
+            }
+        }
+    }
+
+    PropagateVisibility(const core::packages::PackageInfo &package) : package{package} {}
+
+public:
+    // Find uses of export and mark the symbols they mention as exported.
+    ast::ExpressionPtr postTransformSend(core::MutableContext ctx, ast::ExpressionPtr tree) {
+        auto &send = ast::cast_tree_nonnull<ast::Send>(tree);
+        if (send.fun != core::Names::export_()) {
+            return tree;
+        }
+
+        if (send.numPosArgs() != 1) {
+            // an error will have been raised in the packager pass
+            return tree;
+        }
+
+        auto *lit = ast::cast_tree<ast::ConstantLit>(send.getPosArg(0));
+        if (lit == nullptr || lit->symbol == core::Symbols::StubModule()) {
+            // We don't raise an explicit error here, as this is one of two cases:
+            //   1. Export is given a non-constant argument
+            //   2. The argument failed to resolve
+            // In both cases, errors will be raised by previous passes.
+            return tree;
+        }
+
+        auto &db = ctx.state.packageDB();
+        if (lit->symbol.isClassOrModule()) {
+            auto sym = lit->symbol.asClassOrModuleRef();
+            checkExportPackage(ctx, send.loc, db.getPackageNameForFile(sym.data(ctx)->loc().file()));
+            recursiveExportSymbol(ctx, true, sym);
+
+            exportParentNamespace(ctx, sym.data(ctx)->owner);
+        } else if (lit->symbol.isFieldOrStaticField()) {
+            auto sym = lit->symbol.asFieldRef();
+            checkExportPackage(ctx, send.loc, db.getPackageNameForFile(sym.data(ctx)->loc().file()));
+            sym.data(ctx)->flags.isExported = true;
+
+            // When exporting a field, we also export its parent namespace. This is a bit of a hack, and it would be
+            // great to remove this, but this was the behavior of the previous packager implementation.
+            // TODO: can we get away with only marking the owner as exported, and not everything up to the package?
+            exportParentNamespace(ctx, sym.data(ctx)->owner);
+
+            if (sym.data(ctx)->flags.isStaticFieldTypeAlias) {
+                if (auto e = ctx.beginError(send.loc, core::errors::Packager::ExportingTypeAlias)) {
+                    e.setHeader("Exporting a type alias is not allowed in package specifications");
+                    e.addErrorLine(sym.data(ctx)->loc(), "Originally defined here");
+                }
+            }
+        }
```

Comment:
ooc, do we have a test that confirms that it's an error to attempt to export a type member?

Not sure if any of these get close to the problem yet or what but:

```ruby
class A::B < PackageSpec
    export Elem
    export A::B::Elem

    export A::B::Example::Elem
end

class A::B
  extend T::Generic

  Elem = type_member
end

class A::B::Example
  extend T::Generic
  Elem = type_member
end
```

Reply:
We weren't raising an error for this, but I've added one and a test.